File: javascript/leetCodeChallenges/Easy/isomorphic_strings.js
// Given two strings s and t, determine if they are isomorphic.
//
// Two strings s and t are isomorphic if the characters in s can be replaced to get t.
//
// All occurrences of a character must be replaced with another character while preserving the order of characters.
// No two characters may map to the same character, but a character may map to itself.

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
    if (s.length !== t.length)  {
        return false
    }
    const sToT = new Map();
    const tToS = new Map();

    for(let i = 0; i < s.length; i += 1) {
        const c1 = s[i];
        const c2 = t[i];

        if(sToT.has(c1)) {
            if(sToT.get(c1) !== c2) return false
        } else {
            sToT.set(c1, c2)
        };

        if (tToS.has(c2)) {
            if (tToS.get(c2) !== c1) return false;
        } else {
            tToS.set(c2, c1);
        }
    }
    return true;
}

// Time and Space Complexity is O(n)

File: python/LeetCodeChallenges/Easy/isomorphic_strings.py
# Todo Given two strings s and t, determine if they are isomorphic
#  Two strings s and t are isomorphic if the characters in s can be replaced to get t.

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        "egg add"
        s_to_t = {}
        t_to_s = {}

        for c1, c2 in zip(s, t):
            if c1 in s_to_t:
                if s_to_t[c1] != c2:
                    return False
            else:
                s_to_t[c1] = c2

            if c2 in t_to_s:
                if t_to_s[c2] != c1:
                    return False
            else:
                t_to_s[c2] = c1

        return True

# Time and Space complexity is O(n)